arangod/Replication2/ReplicatedLog/Algorithms.cpp

```diff
@@ -0,0 +1,267 @@
+    case ConflictReason::LOG_EMPTY:
+      return "the replicated log is empty";
+    case ConflictReason::LOG_ENTRY_NO_MATCH:
+      return "term mismatch";
+  }
+  LOG_TOPIC("03e11", FATAL, arangodb::Logger::REPLICATION2)
+      << "Invalid ConflictReason "
+      << static_cast<std::underlying_type_t<decltype(r)>>(r);
+  FATAL_ERROR_ABORT();
+}
+
+auto algorithms::detectConflict(replicated_log::InMemoryLog const& log, TermIndexPair prevLog) noexcept
+    -> std::optional<std::pair<ConflictReason, TermIndexPair>> {
+  /*
+   * There are three situations to handle here:
+   *  - We don't have that log entry
+   *    - It is behind our last entry
+   *    - It is before out first entry
+   *  - The term does not match.
+   */
+  auto entry = log.getEntryByIndex(prevLog.index);
+  if (entry.has_value()) {
+    // check if the term matches
+    if (entry->entry().logTerm() != prevLog.term) {
+      auto conflict = std::invoke([&] {
+        if (auto idx = log.getFirstIndexOfTerm(entry->entry().logTerm()); idx.has_value()) {
+          return TermIndexPair{entry->entry().logTerm(), *idx};
+        }
+        return TermIndexPair{};
+      });
+
+      return std::make_pair(ConflictReason::LOG_ENTRY_NO_MATCH, conflict);
+    } else {
+      // No conflict
+      return std::nullopt;
+    }
+  } else {
+    auto lastEntry = log.getLastEntry();
+    if (!lastEntry.has_value()) {
+      // The log is empty, reset to (0, 0)
+      return std::make_pair(ConflictReason::LOG_EMPTY, TermIndexPair{});
+    } else if (prevLog.index > lastEntry->entry().logIndex()) {
+      // the given entry is too far ahead
+      return std::make_pair(ConflictReason::LOG_ENTRY_AFTER_END,
+                            TermIndexPair{lastEntry->entry().logTerm(),
+                                          lastEntry->entry().logIndex() + 1});
+    } else {
+      // this can only happen if we drop log entries, check the code below before removing the assert
+      TRI_ASSERT(false);
+      TRI_ASSERT(prevLog.index < lastEntry->entry().logIndex());
+      TRI_ASSERT(prevLog.index < log.getFirstEntry()->entry().logIndex());
+      // the given index too old, reset to (0, 0)
+      return std::make_pair(ConflictReason::LOG_ENTRY_BEFORE_BEGIN, TermIndexPair{});
+    }
+  }
+}
+
+auto algorithms::updateReplicatedLog(LogActionContext& ctx, ServerID const& serverId,
+                                     RebootId rebootId, LogId logId,
+                                     agency::LogPlanSpecification const* spec) noexcept
+    -> arangodb::Result {
+  return basics::catchToResult([&]() -> Result {
+    if (spec == nullptr) {
+      return ctx.dropReplicatedLog(logId);
+    }
+
+    TRI_ASSERT(logId == spec->id);
+    TRI_ASSERT(spec->currentTerm.has_value());
+    auto& leader = spec->currentTerm->leader;
+    auto log = ctx.ensureReplicatedLog(logId);
+
+    if (leader.has_value() && leader->serverId == serverId && leader->rebootId == rebootId) {
+      auto follower =
```

Comment:
I suggest renaming `follower` to `followers`, as the vector can contain any number of followers.
```suggestion
      auto follower =
```